src/main/java/com/google/api/codegen/config/GapicProductConfig.java

```diff
@@ -485,6 +503,163 @@ private static GapicProductConfig createDummyInstance(
+        if (!Strings.isNullOrEmpty(name.getMethod())) {
+          findAndSetRetry(ib, true, name.getMethod(), rcb.getName(), rpb.getName(), timeout);
+        } else {
+          Interface protoService = protoInterfaces.get(ib.getName());
+          for (Method m : protoService.getMethods()) {
+            findAndSetRetry(ib, false, m.getSimpleName(), rcb.getName(), rpb.getName(), timeout);
+          }
+        }
+      }
+    }
+
+    return builders
+        .keySet()
+        .stream()
+        .map(
+            key -> {
+              return builders.get(key).build();
+            })
+        .collect(Collectors.toList());
+  }
+
+  /** converts a float to a double rounding to the nearest hundredth */
+  private static double floatToDouble(float f) {
+    return Math.round(f * 100) / 100.0;
+  }
+
+  /** adds the retry param and code list objects to the GAPIC interface unless already present */
+  static void addRetryConfigIfAbsent(
+      InterfaceConfigProto.Builder ib,
+      RetryCodesDefinitionProto.Builder rcb,
+      RetryParamsDefinitionProto.Builder rpb) {
+    boolean addRetryCodes = true;
+    for (RetryCodesDefinitionProto c : ib.getRetryCodesDefList()) {
+      if (c.getName().equals(rcb.getName())) {
+        addRetryCodes = false;
+      }
+    }
+
+    if (addRetryCodes) {
+      ib.addRetryCodesDef(rcb);
+    }
+
+    boolean addRetryParams = true;
+    for (RetryParamsDefinitionProto p : ib.getRetryParamsDefList()) {
+      if (p.getName().equals(rpb.getName())) {
+        addRetryParams = false;
+      }
+    }
+
+    if (addRetryParams) {
+      ib.addRetryParamsDef(rpb);
+    }
+  }
+
+  /**
+   * finds the named MethodConfigProto, creates one if absent in the GAPIC interface, and sets retry
+   * config
+   */
+  static void findAndSetRetry(
+      InterfaceConfigProto.Builder ib,
+      boolean overwrite,
+      String method,
+      String rc,
+      String rp,
+      long timeout) {
+    int i = findMethod(ib, method);
+
+    // add a new MethodConfigProto item to the GAPIC interface
+    if (i == -1) {
+      MethodConfigProto.Builder mcp = MethodConfigProto.newBuilder();
+      mcp.setRetryParamsName(rp);
+      mcp.setRetryCodesName(rc);
+      mcp.setName(method);
+      mcp.setTimeoutMillis(timeout);
+      ib.addMethods(mcp);
+      return;
+    }
+
+    MethodConfigProto.Builder mcp = ib.getMethods(i).toBuilder();
+    // don't overwrite a config previously given by gRPC service config
+    if (!overwrite
+        && (mcp.getRetryCodesName().startsWith("retry_policy_")
+            || mcp.getRetryCodesName().equals("no_retry_codes"))) {
+      return;
+    }
+
+    mcp.setRetryParamsName(rp);
+    mcp.setRetryCodesName(rc);
+    mcp.setTimeoutMillis(timeout);
+    ib.setMethods(i, mcp);
+  }
+
+  /** returns the index of the named MethodConfigProto in the GAPIC interface and -1 if not found */
+  static int findMethod(InterfaceConfigProto.Builder service, String method) {
```

Comment:
private?

Reply:
this is "package-private" and I will add the VisibileForTesting annotation so that folks know this is on purpose